### PR TITLE
[CS2113T-T09-2]-nishanthelango-B-DetectAnomalies

### PR DESCRIPTION
**Pull request by nishanthelango**
AddEventCommand now checks if the event being added clashes with any existing events.

### CODE REVIEW

File: src/main/java/duke/command/AddEventCommand.java [hunk 1]
```diff
@@ -16,8 +17,13 @@ public AddEventCommand(List<String> words) {
     @Override
     public void execute(TaskList taskList, Ui ui, Storage storage) throws DukeException {
         Task task = new Event(words.subList(0, words.size()));
-        taskList.add(task);
-        ui.showTaskAdded(taskList.getTasks(), task);
-        storage.save(taskList);
+        try {
+            Event.checkClash((Event) task, taskList);
```
Comment (mohideenik) [inline src/main/java/duke/command/AddEventCommand.java]: No need to typecast. Change `Task task = ... ` above to just `Event task = ...`

File: src/main/java/duke/command/AddEventCommand.java [hunk 2]
```diff
@@ -16,8 +17,13 @@ public AddEventCommand(List<String> words) {
...
+            taskList.add(task);
+            ui.showTaskAdded(taskList.getTasks(), task);
+            storage.save(taskList);
+        } catch (DukeException e) {
```
Comment (mohideenik) [inline src/main/java/duke/command/AddEventCommand.java]: Don't catch here. The `run` function in `Duke.java` does that.

File: src/main/java/duke/task/Event.java [hunk 3]
```diff
@@ -39,10 +39,44 @@ public Event(List<String> input) throws DukeException {
...
     }
 
+    /**
+     * Checks if the event being added clashes with any existing events.
+     *
+     * @param eventToAdd the event task to be added
+     * @param taskList List of tasks
+     * @throws DukeException error if there is a clash
+     */
+    public static void checkClash(Event eventToAdd, TaskList taskList) throws DukeException {
```
Comment (mohideenik) [inline src/main/java/duke/task/Event.java]: For reusability just make this a function that returns a boolean indicating if there is a clash or not. Then, throw the exception inside the `execute` function in the `AddEventCommand` class.

File: src/main/java/duke/task/Event.java [hunk 4]
```diff
@@ -39,10 +39,44 @@ public Event(List<String> input) throws DukeException {
...
+        return eventToAdd.end.after(((Event) task).start) && eventToAdd.end.before(((Event) task).end);
+    }
+
+    private static boolean entireEventClashes(Event eventToAdd, Task task) {
+        return eventToAdd.start.before(((Event) task).start) && eventToAdd.end.after(((Event) task).start);
+    }
+
+    private static boolean isStartOrEndEqual(Event eventToAdd, Task task) {
+        return eventToAdd.start.equals(((Event) task).start) || eventToAdd.end.equals(((Event) task).end);
+    }
```
Comment (mohideenik) [inline src/main/java/duke/task/Event.java]: I think the function signatures of all clash comparison functions here can just be 

`private boolean comparisonFunctionName(Event that)`

Then, the actual clash check should be performed in either `TaskList` as a member function there.

Comment (nishanthelango) [reply src/main/java/duke/task/Event.java]: So should I shift all the clash comparison functions together with the actual clash check function over to TaskList?

Comment (mohideenik) [reply src/main/java/duke/task/Event.java]: No, the function to compare events for clashes should reside within the Event class for encapsulation. Other classes do not need to know how to compare two Event objects. Also, make the clash comparison function non `static`.

Now that we have the comparison function two compare two events (with the signature above), we can invoke it as follows inside another function TaskList.

```
List<Event> events = somehow get only the events
boolean isClashing = events.stream().anyMatch(event -> event.clashesWith(newEvent));
return isClashing
```

Now, the `execute` function can ask the `TaskList` if the new event is gonna clash, then decide to add or throw accordingly.

File: src/main/java/duke/command/AddEventCommand.java (hunk 2 above)
Comment (muserr) [reply src/main/java/duke/command/AddEventCommand.java]: Agree with @mohideenik , the exception will be thrown by the function if needed.

File: src/main/java/duke/task/Event.java (hunk 4 above)
Comment (nishanthelango) [reply src/main/java/duke/task/Event.java]: Alright, made the changes. Thanks!

File: src/test/java/DetectAnomaliesTest.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        assertTrue(taskList.isClashing(event));
+    }
+
+    @Test
+    void isStartOrEndEqual_clash_exceptionThrown() throws DukeException {
+        TaskList taskList = new TaskList();
+        Task task = new Event(getList("Event /at 12/12/2020 1800 to 12/12/2020 1900"));
+        taskList.add(task);
+        Event event = new Event(getList("Event /at 12/12/2020 1800 to 12/12/2020 1830"));
+        assertTrue(taskList.isClashing(event));
```
Comment (mohideenik) [inline src/test/java/DetectAnomaliesTest.java]: Duplicated

**Comment by muserr** [comment]
Looks good! :D

**Pull request merged**